tls/s2n_ktls.c

```diff
@@ -82,38 +104,150 @@ S2N_RESULT s2n_ktls_retrieve_file_descriptor(struct s2n_connection *conn, s2n_kt
+    if (use_server_km) {
+        key = &key_material->server_key;
+        RESULT_GUARD_POSIX(s2n_blob_init(&implicit_iv, conn->server->server_implicit_iv, sizeof(conn->server->server_implicit_iv)));
+        RESULT_GUARD_POSIX(s2n_blob_init(&sequence_number, conn->server->server_sequence_number, sizeof(conn->server->server_sequence_number)));
+    } else {
+        key = &key_material->client_key;
+        RESULT_GUARD_POSIX(s2n_blob_init(&implicit_iv, conn->client->client_implicit_iv, sizeof(conn->client->client_implicit_iv)));
+        RESULT_GUARD_POSIX(s2n_blob_init(&sequence_number, conn->client->client_sequence_number, sizeof(conn->client->client_sequence_number)));
+    }
+
+    /**
+     *= https://www.rfc-editor.org/rfc/rfc4106#section-4
+     *# The salt field is a four-octet value that is assigned at the
+     *# beginning of the security association, and then remains constant
+     *# for the life of the security association.
+     */
+    /* The salt is the first 4 bytes of the IV */
+    RESULT_ENSURE_GTE(implicit_iv.size, TLS_CIPHER_AES_GCM_128_SALT_SIZE);
+    RESULT_CHECKED_MEMCPY(crypto_info->salt, implicit_iv.data, TLS_CIPHER_AES_GCM_128_SALT_SIZE);
+
+    RESULT_ENSURE_GTE(implicit_iv.size, TLS_CIPHER_AES_GCM_128_IV_SIZE);
+    RESULT_CHECKED_MEMCPY(crypto_info->iv, implicit_iv.data, TLS_CIPHER_AES_GCM_128_IV_SIZE);
+
+    RESULT_ENSURE_EQ(sequence_number.size, TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
+    RESULT_CHECKED_MEMCPY(crypto_info->rec_seq, sequence_number.data, TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
+
+    RESULT_ENSURE_EQ(key->size, TLS_CIPHER_AES_GCM_128_KEY_SIZE);
+    RESULT_CHECKED_MEMCPY(crypto_info->key, key->data, TLS_CIPHER_AES_GCM_128_KEY_SIZE);
+
+    return S2N_RESULT_OK;
+}
+#endif
+
+static S2N_RESULT s2n_ktls_set_keys(struct s2n_connection *conn, s2n_ktls_mode ktls_mode,
+        struct s2n_key_material *key_material)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(key_material);
 
     int fd = 0;
-    RESULT_GUARD(s2n_ktls_retrieve_file_descriptor(conn, ktls_mode, &fd));
+    RESULT_GUARD(s2n_ktls_get_file_descriptor(conn, ktls_mode, &fd));
+
+    int tls_tx_rx_mode = 0;
+    RESULT_GUARD(s2n_ktls_get_io_mode(ktls_mode, &tls_tx_rx_mode));
+#if defined(S2N_KTLS_SUPPORTED)
+    /* Only AES_128_GCM for TLS 1.2 is supported at the moment. */
+    struct tls12_crypto_info_aes_gcm_128 crypto_info = { 0 };
+    RESULT_GUARD(s2n_ktls_init_aes128_gcm_crypto_info(conn, ktls_mode, key_material, &crypto_info));
 
-    /* Calls to setsockopt require a real socket, which is not used in unit tests. */
-    if (s2n_in_unit_test()) {
-        return S2N_RESULT_OK;
+    /* Set crypto_info on the socket; enabling kTLS and offloading the TLS
+     * protocol to the kernel. */
+    int ret = s2n_setsockopt(fd, S2N_SOL_TLS, tls_tx_rx_mode, &crypto_info, sizeof(crypto_info));
+    if (ret < 0) {
+        RESULT_BAIL(S2N_ERR_KTLS_ENABLE_CRYPTO);
     }
```

Comment:
I think in this case ENSURE is worse because signage is hard. I would like to keep this as `ret < 0` so that next person (me) doesnt have to rethink the logic of `>=`.